pandas/io/html.py

```diff
@@ -657,12 +647,21 @@ def _parse_raw_thead(self, table):
         thead = table.xpath(expr)
         res = []
         if thead:
-            trs = self._parse_tr(thead[0])
-            for tr in trs:
-                cols = [_remove_whitespace(x.text_content()) for x in
-                        self._parse_td(tr)]
+            # Grab any directly descending table headers first
+            ths = thead[0].xpath('./th')
```

Comment:
Because `_parse_td` with this parser doesn't really differentiate between `td` and `th` elements it was incorrectly parsing headers for things like `spam.html` where `td` and `th` elements are intermixed in the header. Hence to make the parsing more robust and pass the tests, I added an initial search for `th` elements before falling back to the existing behavior.

Even with that I'd argue it's confusing that `_parse_td` is implemented to return `td` and `th` elements and should be refactored to more clearly delineate, but I am trying to minimize behavior change with this PR